[PowerToys Run] Add Support for Uris

royvou opened the pull request:
## Summary of the Pull Request
Tries to resolve the query string as a url and if it is able to parse it as a url will show the url in the searchresultlist.

## References

## PR Checklist
* [x] Applies to #4542
* [x] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/PowerToys) and sign the CLA
* [x] Tests added/passed
* [ ] Requires documentation to be updated
* [x] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #xxx

## Detailed Description of the Pull Request / Additional comments


## Validation Steps Performed

- Added the unit tests for the search results
- Searched for "google.com" and saw "http://google.com/" show up in the results and upon clicking it opens my default browser (Firefox) with the correct url

Commit:
url handler plugin

Commit:
updates

Commit:
Add seperate interface classes
rename to uri module

Commit:
Update path

Commit:
merge master into feature/urlhandler

Commit:
Update implementation to remove slow DNS lookup ( and let browser handle it)

Commit:
tabs to spaces

htcfreek commented:
@royvou 
1. You have missed to change the plugin description.

2. Can we use as plugin icon the icon # EB41 of the font Segeo MDL2 Asset?
_See icon list [here](https://docs.microsoft.com/en-us/windows/uwp/design/style/segoe-ui-symbol-font#icon-list)._



_I know this PR is in draft._

niels9001 commented:
> @royvou
> 
> 1. You have missed to change the plugin description.
> 2. Can we use as plugin icon the icon EB41 of the font Segeo MDL2 Asset?
>    _See icon list [here](https://docs.microsoft.com/en-us/windows/uwp/design/style/segoe-ui-symbol-font#icon-list)._
> 
> _I know this PR is in draft._

I agree, with #4149 we have aligned all plug-in icons by using the Segoe MDL2 Assets icon. 

I have created a version using icon EB41 that uses the right margins and icon size, so you can easily replace the current one :).

![url dark](https://user-images.githubusercontent.com/9866362/88798127-b46b6400-d1a4-11ea-9810-5d59fdc21626.png)

![url light](https://user-images.githubusercontent.com/9866362/88798130-b503fa80-d1a4-11ea-82f6-bf9370fdd0ea.png)

royvou commented:
Great suggestions yes! I will work on these two small improvements and update PR status.

Commit:
merge master in branch

Commit:
- Update icon/assets
- Finalize Project

Commit:
Update wix project

Commit:
Merge remote-tracking branch 'microsoft/master' into feature/urlhandler

royvou commented:
Thanks @niels9001 for the icons, the PR is ready now (I used an icon from the Microsoft Fluent Icon set originally,  (#openInBrowser24Filled at https://github.com/microsoft/fluentui-system-icons/blob/master/icons.md) but now they are from the same icon set!

If anyone reviews this, can they double check if i've done everything right with the wix project (Layout seems correct / installer works correctly but it's my first wix change))

crutkas commented:
moved this for .21, i'll test this right now.

htcfreek commented:
@royvou 
Can you please add an image how the result is look like? For those (including me) who aren't to build and run your code? Thank you.

royvou commented:
![urihandler](https://user-images.githubusercontent.com/16852398/88958215-705d8980-d2a0-11ea-9a27-4da7a1b06fde.png)

I made the image but did not attach it, here is a sample (Where it is the only result) in Dark mode.

htcfreek commented:
I think we should add an description under the uri. Something like "Open in default browser" or as I suggest in the issue to show name and icon of default browser.

htcfreek commented:
Does the URL-Parser support URLs like `bing.com/search?q=gmx`?
Didn't find something similar within the tests.

niels9001 commented:
Nice work, @royvou ! Looking forward to try this out!

Yeah, we either need to have some sort of subtitle (like @htcfreek suggested) or we could update the DataTemplate so that the title is vertically aligned because this looks a bit off. That would be a good thing to do anyways, I'll open a seperate issue for that.

Question: can we easily detect what the default browser is and show its app icon instead of the MDL2 icon?

htcfreek commented:
> Question: can we easily detect what the default browser is and show its app icon instead of the MDL2 icon? 

Setting is saved here:
HKEY_CURRENT_USER\Software\Microsoft\Windows\Shell\Associations\UrlAssociations\http\UserChoice

crutkas commented:
Bottom line should be like "Website" and yeah, if we could pull default browser's icon, that would be baller

looking at the reg value however, any ideas how we backtrack that to an actual app?

niels9001 commented:
Not sure if it works at all, but: https://stackoverflow.com/questions/16863287/how-to-get-icon-of-default-browser-in-c

crutkas commented:
Tried that but pulls IE when my default is Edge Dev branch.  neat code however
![image](https://user-images.githubusercontent.com/1462282/88966272-e215e300-d260-11ea-85de-3f12cdd60475.png)

royvou commented:
Thanks for the suggestions!

- Subtitle -> Will be added, didn't even notice it was missing until you said it.
- Icon for default browser is a great idea! What @Niels suggests seems to retrieve icon for XP, the path @htcfreek mentions is the Vista onwards one (and current).
Seems like HKEY_CURRENT_USER\Software\Microsoft\Windows\Shell\Associations\UrlAssociations\http contains the Id of the app which then can be retreived via HKEY_CLASSES_ROOT\<<ID>>\DefaultIcon (atleast it shows the Path of the .Exe), not sure how to get it for UWP apps yet though (As it contains a ms-resource:// url). Need to investigate a bit for that one.

- regarding urls with querystrings etc those all work as expected, will add a seperate test for that though.

htcfreek commented:
> * Icon for default browser is a great idea! What @niels suggests seems to retrieve icon for XP, the path @htcfreek mentions is the Vista onwards one (and current).
>   Seems like HKEY_CURRENT_USER\Software\Microsoft\Windows\Shell\Associations\UrlAssociations\http contains the Id of the app which then can be retreived via HKEY_CLASSES_ROOT<>\DefaultIcon (atleast it shows the Path of the .Exe), not sure how to get it for UWP apps yet though (As it contains a ms-resource:// url). Need to investigate a bit for that one.

@royvou 
I have found this: [https://stackoverrun.com/de/q/10315156#37432489](https://stackoverrun.com/de/q/10315156#37432489). Don't know if this works.

royvou commented:
SHLoadIndirectString is indeed the solution, but does not take in account the "Theme" of the icon (Looks weird with the "Old" edge), but the Programs plugin contains some logic to retrieve an themed icon.

Will implement these and update the PR.

Commit:
Implement UpdateBrowserIconPath

Commit:
merge master

Commit:
Implemented Microsoft.CodeAnalysis.FxCopAnalyzers

Commit:
Add Language component to installer

Commit:
Update logic to determine icon

royvou commented:
I've updated it with the following logic.

Screenshot of it using the new Edge & Firefox Nightly (in Dark mode)
![DefaultBrowserIcon](https://user-images.githubusercontent.com/16852398/89104433-b12cde00-d419-11ea-8743-2d01c475bc3f.png)

GIF of using it (With copy-paste a url inside it) on light/dark mode with the Old Edge.
![OpenInBrowser](https://user-images.githubusercontent.com/16852398/89104457-d0c40680-d419-11ea-81b6-3ba3f1ed6ba4.gif)

I really like having the icon! Its an option in code to  disable it though, as it is possible it will fall back to the icon if it does not exist.

I've got two items that might need improvement though.
- Should the Browser: Say something browser: Open in Default? Feels a little odd now.
- Currently the default browser icon only changes when the Theme is changed / on startup, so if someone changes it mid-runtime it will still keep the old one (Don't think this is a big issue, so i think this should be acceptable?) But if you are testing you will notice it.

niels9001 commented:
@royvou Oooh, that looks great :)!!!

I'd go for a static "Open in browser". I don't really see the need for mentioning the browser name since we have the browser icon there.

crutkas commented:
First, looks awesome and thanks for helping take the refinements. 

Why it is pulling old edge if you have new edge?

royvou commented:
Well not sure about that one but at set default programs you can still select the old egde (or at least I can?)
It opens the new edge in both cases though! (and you can also select the new edge). Might be a bug. 

(unless people decided to do a side by side install, then it could be either)

Testing the icon out with the old edge icon (when it qa set as the default browser)  was a great way to test UWP programs which were set as defeat browser.

htcfreek commented:
@royvou 
Nice work. I like it. 👍😍

> - Should the Browser: Say something browser: Open in Default? Feels a little odd now.

I think a text like "open in browser" is enough because we have the icon.

> - Currently the default browser icon only changes when the Theme is changed / on startup, so if someone changes it mid-runtime it will still keep the old one (Don't think this is a big issue, so i think this should be acceptable?) But if you are testing you will notice it.

I think this is something we should fix. Maybe in a second PR.

Commit:
Update Translation File to "Open in browser"

royvou commented:
I've updated the title of the PR to more reflect the change(s) done. This PR is getting along !

I've changed the text to "Open in browser", which makes feels a lot better.
![openinbrowser](https://user-images.githubusercontent.com/16852398/89217283-322bd700-d5cc-11ea-825c-4dafb6582fc0.png).

crutkas commented:
1. so awesome.  Love the icon integration.  

2. this happens on http and https ![image](https://user-images.githubusercontent.com/1462282/89355352-06622d00-d670-11ea-986e-215aad453e71.png)

3. @ryanbodrug-microsoft i ran it and looks good, can you give it a once over as well?

royvou commented:
Http is a valid hostname (will probably not resolve to an ip) so yea... This is expected behaviour.

htcfreek commented:
> Http is a valid hostname (will probably not resolve to an ip) so yea... This is expected behaviour. 

Sure. But he didn't type http as host name. He is typing the protocol.

royvou commented:
@htcfreek list makes sense, will add test cases for these and update parser.

Commit:
Added test for typing http://test.com and which result to expect on each keystoke

Commit:
merge master into feature/urlhandler

htcfreek commented on src/modules/launcher/Plugins/Microsoft.Plugin.Uri.UnitTests/UriHelper/ExtendedUriParserTests.cs:
```diff
@@ -0,0 +1,50 @@
+        [TestCase("http", true, "http://http/")]
+        [TestCase("http:", false, null)]
+        [TestCase("http:/", false, null)]
+        [TestCase("http://", false, null)]
```
Is this handled by uri builder itself?

htcfreek commented on src/modules/launcher/Plugins/Microsoft.Plugin.Uri/UriHelper/ExtendedUriParser.cs:
```diff
@@ -0,0 +1,39 @@
+            //Handle common cases UriBuilder does not handle
+            if (input.EndsWith(":", StringComparison.Ordinal)
+                || input.EndsWith(".", StringComparison.Ordinal)
+                || input.EndsWith(":/", StringComparison.Ordinal))
```
Missing test case `://`. Is that case handle by the uri builder?

royvou replied:
Yes. 

There is a test added for it also

https://github.com/royvou/PowerToys/blob/feature/urlhandler/src/modules/launcher/Plugins/Microsoft.Plugin.Uri.UnitTests/UriHelper/ExtendedUriParserTests.cs#L25

htcfreek commented on src/modules/launcher/Plugins/Microsoft.Plugin.Uri.UnitTests/UriHelper/ExtendedUriParserTests.cs:
```diff
@@ -0,0 +1,50 @@
+    {
+        [TestCase("google.com", true, "http://google.com/")]
+        [TestCase("localhost", true, "http://localhost/")]
+        [TestCase("127.0.0.1", true, "http://127.0.0.1/")]
```
Do we need also the following tests?
- 127.0.0.1:80
- [::]
- [::]:80
- [2001:0DB8::1]
- [2001:0DB8::1]:80

somil55 commented on PowerToys.sln:
```diff
@@ -1,602 +1,616 @@
-Microsoft Visual Studio Solution File, Format Version 12.00
```
Can you please fix the git history changes for solution file to only include necessary changes?

crutkas commented:
@royvou, when do you want to try this again?

dsrivastavv reviewed (commented):
The plugin also needs to be added as a project dependency for `PowerLauncher` project.

Btw Nice Work! This would be a great addition to PT Run plugins.

crutkas commented on src/modules/launcher/Plugins/Microsoft.Plugin.Uri/GlobalSuppressions.cs:
```diff
@@ -0,0 +1,47 @@
+﻿// Copyright (c) Microsoft Corporation
```
also need to add in the stylecop reference

```
  <ItemGroup>
    <Compile Include="..\..\..\codeAnalysis\GlobalSuppressions.cs">
      <Link>GlobalSuppressions.cs</Link>
    </Compile>
    <AdditionalFiles Include="..\..\..\codeAnalysis\StyleCop.json">
      <Link>StyleCop.json</Link>
    </AdditionalFiles>
  </ItemGroup>
  <ItemGroup>
    <PackageReference Include="StyleCop.Analyzers">
      <Version>1.1.118</Version>
      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
      <PrivateAssets>all</PrivateAssets>
    </PackageReference>
  </ItemGroup>
```

crutkas replied:
I also think @somil55's escape pattern is correct, mine isn't for the links to globalsuppression and stylecop.json

andrp92 commented:
Can't wait to see this on an upcoming release 🎊

Commit:
Merge remote-tracking branch 'microsoft/master' into feature/urlhandler

Commit:
Implement StyleCop

Commit:
Added ipv6 tests

Commit:
Fix Solution LineBreaks

royvou commented:
All comments are resolved.

Project reference makes sense, but currently no Plugins are direct projectreferences of the powerlauncher project? @somil55

ryanbodrug-microsoft commented:
> 1. so awesome.  Love the icon integration.
> 2. this happens on http and https ![image](https://user-images.githubusercontent.com/1462282/89355352-06622d00-d670-11ea-986e-215aad453e71.png)
> 3. @ryanbodrug-microsoft i ran it and looks good, can you give it a once over as well?

Sorry for the delay.  Taking a look now.

ryanbodrug-microsoft commented on src/modules/launcher/Plugins/Microsoft.Plugin.Uri/Main.cs:
```diff
@@ -0,0 +1,179 @@
+
+        public void Init(PluginInitContext context)
+        {
+            Context = context ?? throw new InvalidOperationException($"{nameof(Context)} should not be NULL");
```
nit:  I think `ArgumentNullException` is more appropriate here.

royvou replied:
I didn't want to throw a NullReferenceException (As per Language spec), this one is indeed more appropriate.

ryanbodrug-microsoft commented:
> All comments are resolved.
> 
> Project reference makes sense, but currently no Plugins are direct projectreferences of the powerlauncher project? @somil55

I think @somil55 is refering to a build dependency so that a developer can simply hit F5 and all dependent projects will build.  You can implement this by right clicking on the project file and selecting the following.  
![image](https://user-images.githubusercontent.com/56318517/89668328-6c2dff00-d892-11ea-9ff8-7e9b4b00f332.png)
![image](https://user-images.githubusercontent.com/56318517/89668435-954e8f80-d892-11ea-9fd2-aad4b6d76cf6.png)

ryanbodrug-microsoft commented on src/modules/launcher/Plugins/Microsoft.Plugin.Uri/Microsoft.Plugin.Uri.csproj:
```diff
@@ -0,0 +1,133 @@
+  
+  <ItemGroup>
+    <PackageReference Include="JetBrains.Annotations" Version="2020.1.0" />
+    <PackageReference Include="Microsoft.CodeAnalysis.FxCopAnalyzers" Version="3.0.0">
```
Awesome.  Thanks for including static analysis! 👍

Commit:
Added Microsoft.Plugin.Uri as build Dependency

ryanbodrug-microsoft commented on src/modules/launcher/Plugins/Microsoft.Plugin.Uri/UriHelper/UriResolver.cs:
```diff
@@ -0,0 +1,16 @@
+{
+    public class UriResolver : IUrlResolver
+    {
+        public bool IsValidHost(System.Uri uri)
```
Does the Uri.TryParse return false if the host name is invalid? If so I think this method is unnecessary.   Otherwise we should probably use or wrap [Uri.CheckHostName](https://docs.microsoft.com/en-us/dotnet/api/system.uri.checkhostname?view=netcore-3.1)

ryanbodrug-microsoft replied:
I think the URI builder will throw if the hostname is invalid and we don't need this check.

royvou replied:
The parses will only contain valid hostnames/ipadresses.

This Resolver was originally used as a DNS lookup check to see if the url is indeed resolvable, but it was decided to let the browser handle this. So this can be removed / or keep it and  look in the future for some additional validation checks (It's a valid hostname, but we don't want the result in the list.

ryanbodrug-microsoft commented on installer/PowerToysSetup/Product.wxs:
```diff
@@ -1013,6 +1018,36 @@
+          <File Id="Uri_$(var.File)" Source="$(var.BinX64Dir)modules\launcher\Plugins\Microsoft.Plugin.Uri\$(var.File)" />
+        <?endforeach?>
+      </Component>
+      <Component Id="UriImagesComponent" Directory="UriImagesFolder" Guid="8C9C1634-28C8-45C4-A8EA-8D4C9B4810D0">
```
@royvou I think you want this to be `UriLanguagesFolder`.  I see the following when running from the installer. 

![image](https://user-images.githubusercontent.com/56318517/89677248-f336a380-d8a1-11ea-9ea0-f75667d2fe1f.png)

ryanbodrug-microsoft replied:
Sorry the comment is regarding this element `<Component Id="UriLanguagesComponent" Directory="UriImagesFolder"`

royvou replied:
I'll double check this, looks like a small find/replace error!

Commit:
Use ArgumentNullException instead of InvalidOperationException

Commit:
Fix wrong Directory in wix installer

ryanbodrug-microsoft reviewed (approved):
Great work!  This is a handy feature. Thanks :)
